### PR TITLE
BUG: resample.agg with read-only data

### LINKED ISSUE #31710: ValueError: buffer source array is read-only
#### Code Sample, a copy-pastable example if possible

```python
>>> index = pd.date_range('2020', 'now', freq='1h') 
>>> arr = np.zeros_like(index) 
>>> arr.setflags(write=False)
>>> pd.Series(arr, index=index).resample('1d').agg('last')

-------------------------------------------------------------------------
~/pandas/pandas/_libs/groupby.pyx in pandas._libs.groupby.group_last()
~/pandas/pandas/_libs/groupby.cpython-37m-x86_64-linux-gnu.so in View.MemoryView.memoryview_cwrapper()
~/pandas/pandas/_libs/groupby.cpython-37m-x86_64-linux-gnu.so in View.MemoryView.memoryview.__cinit__()
ValueError: buffer source array is read-only
```
#### Problem description

Groupby fails on some read-only buffers (I couldn't quickly reproduce it with `.groupby()` itself, sorry).

The prime solution would be to add `const` specifier to the input `values` here (and related entries):
https://github.com/pandas-dev/pandas/blob/0b6debf825db39f8f63d300cb8007a7996d002fc/pandas/_libs/groupby.pyx#L853

if it were not for Cython's non-support of const fused types (https://github.com/cython/cython/issues/1772), resolved in (https://github.com/cython/cython/pull/3118), but despite [miniscule change](https://github.com/cython/cython/pull/3118/files?file-filters%5B%5D=.py) only scheduled for release in Cython 3.0. I guess wait until then.

#### Expected Output

Resampling/groupby works with read-only arrays.

#### Output of ``pd.show_versions()``

pandas 1.1.0.dev0+361.gf0b00f887
cython 0.29.14

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #31710
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -840,11 +840,15 @@ def _slice(self, slobj: slice, axis: int = 0) -> "Series":
...
+        key = lib.item_from_zerodim(key)
 
         if key is Ellipsis:
             return self
 
-        key_is_scalar = is_scalar(key)
+        # check for is_list_like/slice instead of is_scalar to allow non-standard
+        #  scalars through, e.g. cftime.datetime needed by xarray
+        #  https://github.com/pydata/xarray/issues/3751
+        key_is_scalar = not is_list_like(key) and not isinstance(key, slice)
```
Comment (jorisvandenbossche) [inline pandas/core/series.py]: This sneaked in from a different issue / pr?

Comment (jbrockmendel) [reply pandas/core/series.py]: looks like it, will revert, thanks

File: pandas/_libs/groupby.pyx [hunk 2]
```diff
@@ -848,11 +848,13 @@ cdef inline bint _treat_as_na(rank_t val, bint is_datetimelike) nogil:
...
 
 
+# GH#31710 use memorviews once cython 0.30 is released so we can
+#  use `const rank_t[:, :] values`
 @cython.wraparound(False)
 @cython.boundscheck(False)
 def group_last(rank_t[:, :] out,
                int64_t[:] counts,
-               rank_t[:, :] values,
+               ndarray[rank_t, ndim=2] values,
```
Comment (WillAyd) [inline pandas/_libs/groupby.pyx]: Does this limitation really apply to any place where we are using memoryviews? Or just isolated to this?

Comment (jbrockmendel) [reply pandas/_libs/groupby.pyx]: my understanding is that it applies to memoryviews with fused types

**Comment by jreback** [comment]
thanks @jbrockmendel

**Comment by WillAyd** [comment]
Just a heads up I think this introduced some new build warnings. I've had a separate PR to clean those up and error going forward so can fall back to that if required, but if you have time to look in the interim would be appreciated as well

Here is what I see

```sh
pandas/_libs/groupby.c:37458:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:38233:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:39008:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:39781:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:40563:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:41999:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:42820:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:43641:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:44460:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:45288:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/hashtable.c:63216:20: warning: unused function '__pyx_memview_get_object' [-Wunused-function]
  static PyObject *__pyx_memview_get_object(const char *itemp) {
                   ^
pandas/_libs/hashtable.c:63221:12: warning: unused function '__pyx_memview_set_object' [-Wunused-function]
static int __pyx_memview_set_object(const char *itemp, PyObject *obj) {
           ^
pandas/_libs/groupby.c:55168:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:55970:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:56772:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:57568:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:58985:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:59784:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:60583:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
pandas/_libs/groupby.c:61376:30: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and
      'size_t' (aka 'unsigned long') [-Wsign-compare]
  __pyx_t_3 = ((!((__pyx_t_2 == __pyx_t_1) != 0)) != 0);
                   ~~~~~~~~~ ^  ~~~~~~~~~
```

**Comment by jbrockmendel** [comment]
@WillAyd where does this sit on the annoyance-problem scale?  i.e. is waiting for cython 0.30 a viable option?

**Pull request merged**